### PR TITLE
Faster block matching to improve overall speed

### PR DESCRIPTION
**Pull request by KedoKudo**
This PR introduces the following changes:

- Add contributing guide
- Migrate patch grouping to numba to improve speed
- Migrate hyper block generation to numba to improve speed

## Test instructions

- Checkout branch and update/create conda environment.
- Run the unit test.
- Run the example notebook in notebooks.

You should see that the closed source `bm3d` is significantly slower than `bm3dornl` implementation. On ornl's CADES, the difference is:
- `bm3d`: ~1 min 9s per sinogram
- `bm3dornl`: ~ 8s per sinogram

> EWM item: [5225](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=5225)

### CODE REVIEW

**Review by jmborr — COMMENTED** [review]
the notebook runs for me

File: src/bm3dornl/block_matching.py [hunk 1]
```diff
@@ -4,10 +4,10 @@
 import numpy as np
 from typing import Tuple, Optional
 from bm3dornl.utils import (
+    compute_hyper_block,
+    compute_signal_blocks_matrix,
+    get_patch_numba,
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply src/bm3dornl/block_matching.py:9]: unit test is added

File: src/bm3dornl/block_matching.py [hunk 2]
```diff
@@ -4,10 +4,10 @@
 import numpy as np
 from typing import Tuple, Optional
 from bm3dornl.utils import (
+    compute_hyper_block,
+    compute_signal_blocks_matrix,
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply src/bm3dornl/block_matching.py:8]: unit test is added

File: src/bm3dornl/block_matching.py [hunk 3]
```diff
@@ -150,35 +134,14 @@ def get_hyper_block(
...
-
-        TODO:
-        -----
-        - use multi-processing to further improve the speed of block building
         """
-        group_size = len(self.signal_blocks_matrix)
-        block = np.empty(
-            (group_size, num_patches_per_group, *self.patch_size), dtype=np.float32
+        source_image = self._image if alternative_source is None else alternative_source
+        block, positions = compute_hyper_block(
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply src/bm3dornl/block_matching.py:139]: unit test added

**Pull request merged**